### PR TITLE
Add tests to confirm detection of various integer types

### PR DESCRIPTION
**Pull request by ajnelson-nist**
This patch series will add tests to confirm behaviors of XSD integer types aside from `xsd:integer`.  It stems from trying to test a sample with a literal `"0"^^xsd:positiveInteger` and not receiving an error.

It is not entirely clear to me if corrections from this patch series should be in pySHACL or upstream in RDFLib.

### CODE REVIEW

**Comment by ajnelson-nist** [comment]
@ashleysommer , I don't intend for this to block your plans to do a release soon.

**Comment by ajnelson-nist** [comment]
It looks like discussion of how to resolve this PR should be anchored upstream, on [RDFLib Issue 1757](https://github.com/RDFLib/rdflib/issues/1757).

**Comment by ajnelson-nist** [comment]
@ashleysommer , I stumbled across a data XPASS today that reminded me about this open issue.

It looks like there is still a pause on this while waiting for the next `rdflib` release, which hopefully incorporates [PR 2003](https://github.com/RDFLib/rdflib/pull/2003).  However, I'd like to be ready for the future, and have a typing question for you.

I'm trying to define a shape on a property that has `rdfs:range xsd:positiveInteger`.  Which would be the correct syntax for this?

What I have now is the following, which doesn't throw a validation error on passing in `"0"^^xsd:integer`:

```turtle
ex:my-shape
  a sh:PropertyShape ;
  sh:datatype xsd:positiveInteger ;
  sh:nodeKind sh:Literal ;
  sh:path ex:my-property ;
  sh:targetSubjectsOf ex:my-property ;
  .
```

I figure I can get the important bit about needing to be >=1 across with `sh:minInclusive`, while waiting for the datatype-based enforcement.  But, what is the syntax that should be used?

```turtle
ex:my-shape sh:minInclusive "1"^^xsd:positiveInteger .
```

Or:

```turtle
ex:my-shape sh:minInclusive "1"^^xsd:integer .
```

**Comment by ashleysommer** [comment]
Hi @ajnelson-nist 

Sorry, I forgot about this PR, and I didn't see your typing question above.

You're right, to emulate `rdfs:positiveInteger` in older RDFLib versions, you will need
`sh:datatype xsd:positiveInteger ; sh:minInclusive "1"^^xsd:integer ;`

Note, in Turtle, numerals are implicitly `xsd:integer` so for minInclusive you can simply use sh:minInclusive 1;`

HOWEVER, RDFlib v6.2.0 has been out for a few weeks, and has the new ill-typed literals feature. I have just released a new version of PySHACL (v0.20.0) that takes advantage of this feature when evaluating `sh:datatype` that allows PySHACL to leverage RDFLib's more comprehensive data type checking (sorry, I forgot you had this PR open to do the same).

Note also, RDFLib v6.2.0 has added a new check for `xsd:positiveInteger` among others, so I believe if you upgrade your RDFLib to v6.2.0 and PySHACL to v0.20.0, your issue with type checking should now be resolved.

In regard to the release candidate, that was a little premature. v0.19.x is no longer a release candidate, some big changes were implemented with the move to RDFLib v6.2.0, so the new PySHACL v0.20.0 is the new release candidate.

**Comment by ajnelson-nist** [comment]
Hi @ashleysommer ,

Apologies for the delayed reply, I was on annual leave.

Thanks for integrating the ill-typed literals feature.  Unfortunately, I'm not quite sure how to use it.  I would still like this PR to be eventually be merged because the unit test will serve as a demonstration of triggering and detecting the errors.  I think the test still needs a little more development to take care of what I'm not clear on.

I believe *something* between `pyshacl` and `rdflib` should raise an error for `"-1"^^xsd:positiveInteger`, and the other more numeric samples in the unit test.  It's not clear to me if this is supposed to cause an effect in the SHACL validation results graph.  I would like it to, but I don't know if the SHACL specification covers datatype constraints and ill-typed literals, so I don't know if it's a bend of the SHACL specification to have `"-1"^^xsd:positiveInteger` fail a `sh:datatype xsd:positiveInteger` constraint.  I'd personally prefer this, because I think that's at least in the spirit of SHACL, and possibly in scope of SHACL by vaguely pointing at how `sh:pattern` gives us regexes.

File: test/test_integers.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        (EX_ONT.propNonPositiveInteger, Literal("1", datatype=XSD.nonPositiveInteger)),
+        (EX_ONT.propPositiveInteger, Literal("-1", datatype=XSD.positiveInteger)),
+        (EX_ONT.propPositiveInteger, Literal("0", datatype=XSD.positiveInteger)),
+    }
+    computed_path_values: Set[Tuple[URIRef, Literal]] = set()
+
+    # .triples() is used somewhat redundantly instead of using SPARQL, to avoid the possibility of any extra interpretation steps with SPARQL conversion.
+    for triple0 in graph.triples((None, RDF.type, SH.ValidationResult)):
+        n_validation_result = triple0[0]
+        for triple1 in graph.triples((n_validation_result, SH.path, None)):
```
Comment (ashleysommer) [inline test/test_integers.py:221]:
```suggestion
        for triple1 in graph.triples((n_validation_result, SH.resultPath, None)):
```

**Comment by ashleysommer** [comment]
Hi @ajnelson-nist 

>  Unfortunately, I'm not quite sure how to use it
> I believe something between pyshacl and rdflib should raise an error for "-1"^^xsd:positiveInteger,

I'm not sure I'm following. What kind of error are you expecting?

As-of PySHACL v0.20.0 (and RDFLib v6.2.0) the units tests you contributed now pass. PySHACL now detects the ill-typed literals when executing the `sh.datatype` constraint, and that constraint emits a validation result as expected.

> I don't know if it's a bend of the SHACL specification to have `"-1"^^xsd:positiveInteger` fail a `sh:datatype xsd:positiveInteger` constraint

It is in the [textual definition of the `sh:datatype` specification](https://www.w3.org/TR/shacl/#DatatypeConstraintComponent) in the [SHACL Spec document](https://www.w3.org/TR/shacl/#DatatypeConstraintComponent)

> "A [literal](https://www.w3.org/TR/shacl/#dfn-literal) matches a datatype if the [literal](https://www.w3.org/TR/shacl/#dfn-literal)'s datatype has the same [IRI](https://www.w3.org/TR/shacl/#dfn-iri) and, for the datatypes supported by SPARQL 1.1, **_is not an [ill-typed](https://www.w3.org/TR/rdf11-concepts#section-Graph-Literal) literal_**."

Before RDFLIb v6.2.0 was released, PySHACL had no way to determine if a Literal was ill-typed (beyond some crude string, int and float checks). That is why it these tests were not able to pass before now. But since RDFLib v6.2.0 was released, PySHACL can now take advantage of the new ill-typed Literal test and incorporates that check into the `sh:datatype` constraint validation algorithm.

Thank you for your contribution of this unit test.

**Comment by ajnelson-nist** [comment]
Huh.  I'm not sure why that test was still failing on my laptop yesterday, but it does indeed pass on my laptop now.  Apologies for the confusion.  Thank you for working through this with me, and for the spec. citation!

**Pull request merged**